python/idsse_common/idsse/common/grid_proj.py

```diff
@@ -0,0 +1,86 @@
+
+
+class GridProj:
+    """Wrapper for pyproj instance with methods to convert to and from geographic coordinates, pixels, etc."""
+    def __init__(self,
+                 crs: CRS,
+                 latitude: float,
+                 longitude: float,
+                 width: float,
+                 height: float,
+                 dx: float,
+                 dy: Optional[float] = None):
+        # pylint: disable=too-many-arguments,unpacking-non-sequence
+        self._trans = Transformer.from_crs(crs.geodetic_crs, crs)
+        self._x_offset, self._y_offset = self._trans.transform(longitude, latitude)
+        self._w = width
+        self._h = height
+        self._dx = dx
+        self._dy = dy if dy else dx
+
+    @classmethod
+    def from_proj_grid_spec(cls, proj_string: str, grid_string: str) -> Self:
+        """ Create GridProj instance from projection grid specs
+
+        Args:
+            proj_string (str): pyproj projection string
+            grid_string (str): Grid string
+
+        Returns:
+            Self: New instance of GridProj which can then be converted to any format
+        """
+        crs = CRS.from_proj4(proj_string)
+        grid_args = {
+            key[1:]: float(value)
+            for key, value in (
+                pair.split('=') for pair in grid_string.split(' ')
+            )
+        }
+        return GridProj(crs,
+                        grid_args['lat_ll'], grid_args['lon_ll'],
+                        int(grid_args['w']), int(grid_args['h']),
+                        grid_args['dx'], grid_args['dy'])
+
+    def map_proj_to_pixel(self, x, y) -> Tuple[float, float]:
+        """Map projection to a pixel"""
+        return self.map_geo_to_pixel(*self._trans.transform(x, y))  # pylint: disable=not-an-iterable
+
+    def map_pixel_to_proj(self, x: float, y: float) -> Tuple[Any, Any]:
+        """Map pixel to a projection"""
+        return self._trans.transform(*self.map_pixel_to_geo(x, y), direction=TransformDirection.INVERSE)
+
+    def map_proj_to_geo(self, x, y) -> Tuple[float, float]:
+        """Map projection to geographic coordinates"""
+        return self._trans.transform(x, y)
+
+    def map_pixel_to_geo(self, x: float, y: float) -> Tuple[float, float]:
+        """Map pixel to geographic coordinates"""
+        return x * self._dx + self._x_offset, y * self._dy + self._y_offset
+
+    def map_geo_to_proj(self, x: float, y: float) -> Tuple[Any, Any]:
+        """Map geographic coordinates to a projection"""
+        return self._trans.transform(x, y, direction=TransformDirection.INVERSE)
+
+    def map_geo_to_pixel(self, x: float, y: float) -> Tuple[int, int]:
+        """Map geographic coordinates to a pixel
+        
+        Pixels are only identified by whole numbers when graphically rendered, 
+        so the transformed numerical values (floats) can be safely rounded to ints
+        """
+        return round((x - self._x_offset) / self._dx), round((y - self._y_offset) / self._dy)
```

Comment:
@Geary-Layne it's safe to assume that pixels can be rounded integers, correct? From my general understanding of how browsers actually paint onto a display, a fraction of a pixel is not possible.

Reply:
I ask because unit testing uncovered this strange precision behavior, before I made the change here to round the resulting pixel. I don't think this extra precision is meaningful (or comprehendible) for a client:

```python
geo_x, geo_y = self.map_pixel_to_geo(0, 1)
pixel_x, pixel_y = self.map_geo_to_pixel(geo_x, geo_y)

print(pixel_x, pixel_y)  # (0.0 0.9999999999951331)
```

Reply:
You are correct that for some uses, an non integer pixel value is not meaningful and is problematic. There are however some use case for this code (now that it is a general tool and not specific to the slice operator) where you would want floating point numbers. Also, sometime the use would dictate round and sometime floor.
 
In you test code the precision error is totally executable, and you should us approx().

When I'd implement code like this class in Java and python in the past, I supported an option argument with _to_pixel() function. Something like:
def map_geo_to_pixel(self, x: float, y: float, round: enum=None)
where round can be None, Round, Floor.
Not sure the best way to do this in python, I think I used magic strings before which isn't the best way, and actually making enum seems excessive.

Reply:
I don't feel like Enum is overkill here. It makes the comparison easy, although I didn't find an elegant way to directly map, and invoke, Enum members to functions. 

Simple if statements seemed sufficient for now, and default to no rounding as previously defined in the data-access-service version of GridProj.